File: 021Activitiesandintents/app/src/main/java/com/example/a021activitiesandintents/MainActivity.java
package com.example.a021activitiesandintents;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {
    public static final String EXTRA_MESSAGE  =
            "com.example.android.twoactivities.extra.MESSAGE";
    private EditText mMessageEditText;
    private static final String LOG_TAG =
            MainActivity.class.getSimpleName();
    public static final int TEXT_REQUEST = 1;
    private TextView mReplyHeadTextView;
    private TextView mReplyTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mMessageEditText = findViewById(R.id.editText_main);
        mReplyHeadTextView = findViewById(R.id.text_header_reply);
        mReplyTextView = findViewById(R.id.text_message_reply);
    }

    @Override//요청 코드, 결과 코드(RESULT_OK or Cancled), 인텐트 데이터를 입력값으로 받음
    public void onActivityResult(int requestCode,
                                 int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode == TEXT_REQUEST){//텍스트 리퀘스트 값과 요청코드가 같고
            if(resultCode == RESULT_OK){//결과 코드가 RESULT_OK라면
                String reply = data.getStringExtra(SecondActivity.EXTRA_REPLY);
                //SecondActivity.EXTRA_REPLY의 값(reply)를 String형태의 데이터로 reply에 입력
                mReplyHeadTextView.setVisibility(View.VISIBLE);
                mReplyTextView.setText(reply);
                mReplyTextView.setVisibility(View.VISIBLE);
                //그후 헤더 가시성을 visible, 텍스트 내용을 reply로 입력 후 가시성 visible로 변경
            }

        }
    }

    public void launchSecondActivity(View view) {
        Log.d(LOG_TAG, "Button clicked!");
        Intent intent = new Intent(this, SecondActivity.class);
        String message = mMessageEditText.getText().toString();//mMessageEditTExt의 텍스트를 message에 입력
        intent.putExtra(EXTRA_MESSAGE, message);//인텐트에 EXTRA...를 키로 message를 값으로 추가
        startActivityForResult(intent,TEXT_REQUEST);

    }
}

File: 01LinearLayoutusingtheLayoutEditor/app/src/main/java/com/example/a01linearlayoutusingthelayouteditor/MainActivity.kt
package com.example.a01linearlayoutusingthelayouteditor


import android.content.Context
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.done_button).setOnClickListener {
            addNickname(it)
        }
        findViewById<TextView>(R.id.nickname_text).setOnClickListener {
            updateNickname(it)
        }
    }

    //버튼을 누르면
    private fun addNickname(view: View){//view = done_button
        val editText = findViewById<EditText>(R.id.nickname_edit)//editText에 nickname_edit값 삽입
        val nicknameTextView = findViewById<TextView>(R.id.nickname_text)
        nicknameTextView.text = editText.text//nicknameTextBiew = 입력한 editText의 text값
        editText.visibility = View.GONE //editText의 가시성 -> gone
        view.visibility = View.GONE//view = done_button의 가시성 -> gone
        nicknameTextView.visibility = View.VISIBLE//nicknameTextView의 가시성 -> visible

        val inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager //InputMethodManager타입으로
        //getSystemService = 관리 매니저 객체 반영,INPUT_METHOD_SERVICE 문자 입력 메소드에 접근하기 위해 같이 사용
        inputMethodManager.hideSoftInputFromWindow(view.windowToken, 0)
        //hideSoftInputFromWindow = 입력창을 가리기 위한 요청
    }
    /*
    즉 처음 nickname_edit 레이아웃에 이름을 입력후 버튼을 클릭하면
    editText의 nickname_edit값을 넣은 후 nicknameTextView 프로퍼티엔 nickname_text id값을 삽입
    그 후 nickanameTextView의 텍스트 값(nickname_text.text)에 editText의 텍스트값을 입력한 후
    editText와, view(done_button)의 가시성을 gone으로
    nickanameTextView(nickname_text.visibility)의 가시성을 Visibility로 변경하여 텍스트를 보여줌
    */
    private fun updateNickname(view: View){
        val editText = findViewById<EditText>(R.id.nickname_edit)
        val doneButton = findViewById<TextView>(R.id.done_button)

        editText.visibility = View.VISIBLE
        doneButton.visibility = View.VISIBLE
        view.visibility = View.GONE
        
        editText.requestFocus()//nickname_text클릭시 editText에 현재 포커스를 받는것을 알려줌, 커서가 나타남

        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.showSoftInput(editText, 0)
        //nickname_text 클릭시 실행시 키보드 나타나게
    }
    /*
    위와 비슷, 반대로 화면에 보이는 nickname_text를 클릭하게 되면 숨겨놓았던
    editText와 doneButton의 가시성을 visibility로 변경
    텍스트 변경후 doneButton을 누르면 다시시 addNickname(view:View)가 실행
    */
}

